File: samples-java/src/main/java/io/github/ilaborie/loops/samples/montecarlo/MonteCarlo.java
package io.github.ilaborie.loops.samples.montecarlo;

import java.text.NumberFormat;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static io.github.ilaborie.loops.samples.montecarlo.Point.compute;
import static io.github.ilaborie.loops.samples.montecarlo.Point.newPoint;

public final class MonteCarlo {

    private MonteCarlo() {
        throw new AssertionError("private");
    }

    public static void main(String[] args) {
        NumberFormat format = NumberFormat.getIntegerInstance();
        IntStream.rangeClosed(2, 10)
                .map(e -> (int) Math.pow(10, e))
                .peek(count -> System.out.printf("Compute with %s ...%n", format.format(count)))
                .mapToDouble(MonteCarlo::monteCarloStream)
                .forEach(pi -> System.out.println("Result: " + pi));
    }

    // For Loop
    public static double monteCarloFor(int count) {
        int inCircle = 0;
        for (int i = 0; i < count; i++) {
            Point p = newPoint();

            if (p.inCircle()) {
                inCircle++;
            }
        }
        return compute(count, inCircle);
    }

    // Recursion
    public static double monteCarloRecursion(int count) {
        int inCircle = monteCarloRecursionAux(count, 0);
        return compute(count, inCircle);
    }

    private static int monteCarloRecursionAux(int count, int inCircle) {
        if (count == 0) {
            return inCircle;
        }
        Point p = newPoint();

        if (p.inCircle()) {
            return monteCarloRecursionAux(count - 1, inCircle + 1);
        } else {
            return monteCarloRecursionAux(count - 1, inCircle);
        }
    }

    // Stream
    public static double monteCarloStream(int count) {
        int inCircle = (int) Stream.generate(Point::newPoint)
                .limit(count)
                .filter(Point::inCircle)
                .count();

        return compute(count, inCircle);
    }


    public static double monteCarloStreamParallel(int count) {
        int inCircle = (int) Stream.generate(Point::newPoint)
                .unordered()
                .parallel()
                .limit(count)
                .filter(Point::inCircle)
                .count();

        return compute(count, inCircle);
    }


    public static double monteCarloStreamParallel2(int count) {
        int inCircle = (int) Stream.generate(Point::newPoint2)
                .unordered()
                .parallel()
                .limit(count)
                .filter(Point::inCircle)
                .count();

        return compute(count, inCircle);
    }


}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/montecarlo/monteCarlo.kt
package io.github.ilaborie.loops.samples.montecarlo

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.runBlocking
import java.text.NumberFormat


fun main() {
    val format = NumberFormat.getIntegerInstance()
    (2..20)
        .map { e -> Math.pow(10.0, e.toDouble()).toInt() }
        .onEach { count -> System.out.printf("Compute with %s ...%n", format.format(count.toLong())) }
        .map { monteCarloSequence(it) }
        .forEach { pi -> println("Result: $pi") }
}

// For Loop
fun monteCarloFor(count: Int): Double {
    var inCircle = 0
    for (i in 0 until count) {
        val p = newPoint()

        if (p.inCircle()) {
            inCircle++
        }
    }
    return compute(count, inCircle)
}

// Recursion
fun monteCarloRecursion(count: Int): Double {
    tailrec fun aux(count: Int, inCircle: Int): Int =
        if (count == 0) inCircle
        else {
            val p = newPoint()
            aux(count - 1, if (p.inCircle()) inCircle + 1 else inCircle)
        }

    val inCircle = aux(count, 0)
    return compute(count, inCircle)
}

// Collections
fun monteCarloCollection(count: Int): Double {
    val inCircle = (1..count)
        .map { newPoint() }
        .count { it.inCircle() }

    return compute(count, inCircle)
}

// Sequences (collections lazy)
fun monteCarloSequence(count: Int): Double {
    val inCircle = generateSequence { newPoint() }
        .take(count)
        .count { it.inCircle() }

    return compute(count, inCircle)
}

fun monteCarloSequenceParallel(count: Int): Double {
    val inCircle = sequence {
        yieldAll(generateSequence { newPoint() })
    }
        .take(count)
        .count { it.inCircle() }

    return compute(count, inCircle)
}

fun monteCarloSequenceParallel2(count: Int): Double {
    val inCircle = sequence {
        yieldAll(generateSequence { newPoint2() })
    }
        .take(count)
        .count { it.inCircle() }

    return compute(count, inCircle)
}

fun monteCarloSequenceParallelOther(count: Int): Double {
    val inCircle = sequence {
        yieldAll(generateSequence { newPoint2() })
    }
        .take(count)
        .count {
            //            println("Thread ${Thread.currentThread().name}")
            it.inCircle()
        }

    return compute(count, inCircle)
}

fun monteCarloSequenceParallelOtherAlex(count: Int): Double {
    val inCircle = (1..count).toList()
        .parallelStream()
        .map { newPoint() }
        .filter {
            //            println("Thread ${Thread.currentThread().name}")
            it.inCircle()
        }
        .count()

    return compute(count, inCircle.toInt())
}

// Flow
@ExperimentalCoroutinesApi
fun monteCarloFlow(count: Int): Double {
    val inCircle = runBlocking {
        flow {
            while (true) {
                emit(newPoint())
            }
        }
            .take(count)
            .count { it.inCircle() }
    }

    return compute(count, inCircle)
}
